homeassistant/components/elgato/config_flow.py

```diff
@@ -41,10 +43,12 @@ async def async_step_user(
 
         return self._async_create_entry()
 
-    async def async_step_zeroconf(self, discovery_info: dict[str, Any]) -> FlowResult:
+    async def async_step_zeroconf(
+        self, discovery_info: DiscoveryInfoType
```

Comment:
Or maybe it would it be better to use `zeroconf.HaServiceInfo` ?
```suggestion
        self, discovery_info: zeroconf.HaServiceInfo
```

Reply:
Actually, it’s `unsafe to override a method with a more specific argument type, as it violates the Liskov substitution principle. For return types, it’s unsafe to override a method with a more general return type.`

Reply:
I think we should adjust the upstream method typing for this in a second-round (once all integrations have been adjusted to `ZeroconfServiceInfo`).

Reply:
If we want to go ahead with this, why do it in a second pass? If @bdraco agrees, I can adjust `async_step_zeroconf` right away in `config_entries.py`. I think we just need to add a cast from `ZeroconfServiceInfo` to `DiscoveryInfoType` in the default implementation:
```python
    async def async_step_zeroconf(
        self, discovery_info: ZeroconfServiceInfo
    ) -> data_entry_flow.FlowResult:
        """Handle a flow initialized by Zeroconf discovery."""
        return await self.async_step_discovery(cast(DiscoveryInfoType, discovery_info))
```

Reply:
I've just run some extra tests, and I think it might be easier to first change the result type to `ZeroconfServiceInfo` in `config_entries.py`, and then finish the "pass" on all the components.
If we do it now, then less code to change to satisfy `mypy`

I'll prepare a separate PR for this.

Reply:
PR created #59503

Reply:
Rebased following the merge of PR 59503